### ISSUE TITLE
Jinja2 attr() filter not working in ansible

### ISSUE THREAD

**Issue by GoGoris** [ansible/ansible#24065]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
jinja2

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = ...
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### CONFIGURATION
```properties
[defaults]
host_key_checking = False
retry_files_enabled = False
gather_facts = smart
```

##### OS / ENVIRONMENT
CentOS 7

##### SUMMARY
I noticed some very odd behaviour. I can't get the jinja2 attr() filter to work with ansible. 
The error message says that the attribute is not present in the object but I can access the attribute perfectly fine when using another syntax.

###### Context
```yaml
  - name: copy jar to remote
    copy:
      src: "{{ jar }}"
      dest: "/tmp/test"
      mode: 0700
    register: copy_jars
    with_fileglob:
      - "*.jar"
    loop_control:
      loop_var: jar
```
###### Failing:
```yaml
  - debug:
      msg: "{{ copy_jars.results | first | attr('dest') }}"
```

###### Working:
```yaml
  - debug:
      msg: "{{ copy_jars.results[0].dest }}"
```
###### Working:
```yaml
  - set_fact:
      tempdir: "{{ copy_jars.results | first }}"

  - debug:
      msg: "{{ tempdir.dest }}"
```

##### EXPECTED RESULTS
```json
ok: [api] => {
    "changed": false,
    "msg": "/tmp/tmp.sP3L4bbFtv/auth-server-test.jar"
}
```

##### ACTUAL RESULTS
```json
fatal: [api]: FAILED! => {
    "failed": true,
    "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'dict object' has no attribute 'dest'\n\nThe error appear
s to have been in '/mnt/Projects/internal-auth-server/ansible/roles/ing.lease.api-server/tasks/deploy.yml': line 40, column 5, but may\nbe elsewhere in the file depending on the
exact syntax problem.\n\nThe offending line appears to be:\n\n\n  - debug:\n    ^ here\n"
}
```

**Comment by asouchang** [comment]
Perhaps it's caused by the syntax error in the template.
```
"{{ copy_jars.results | first | attr{'dest') }}"
                                    ^
                                    Here
```

**Comment by GoGoris** [comment]
Sorry this was a typo in the bug report, it is correct in the code. So this is not the issue.

**Comment by jctanner** [comment]
Confirmed behavior. After some debugging, I think this is a jinja problem. Still digging in ...

**Comment by jctanner** [comment]
This is the "root" function jinja compiled for the expression ...

```
def root(context, environment=environment):
    l_copyres = context.resolve('copyres')
    t_1 = environment.filters['attr']
    t_2 = environment.filters['first']
    if 0: yield None
    yield to_string(
        environment.finalize(
            t_1(
                environment,
                t_2(
                    environment,
                    environment.getattr(
                        l_copyres,
                        'results'
                    )
                 ), 
                 'dest'
             )
         )
    )
```

t_1 is "attr", who's source can be found here:
https://github.com/pallets/jinja/blob/master/jinja2/filters.py#L801-L823

t_2 is "first", who's source can be found here:
https://github.com/pallets/jinja/blob/master/jinja2/filters.py#L345-L350

**Comment by jctanner** [comment]
Jinja's attr() filter is basically a python getattr(). That function is useful for fetching an "attribute" of an object. The keys and values in a dictionary can not be accessed via getattr because they are not attributes. A dictonary's attributes are:

['clear', 'copy', 'fromkeys', 'get', 'has_key', 'items', 'iteritems', 'iterkeys', 'itervalues', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values', 'viewitems', 'viewkeys', 'viewvalues']

The error message in the description is telling you that the dictionary doesn't have a 'dest' "attribute", which is the correct error. You need to access the 'dest' key in a different way.

To get a specific value in a dictionary by the key name, the get() function/attribute/method is used. The normal way to do that in jinja is "{{ foo['key'] }}" or "{{ foo.key }}". However, you are using the first filter so you can't just do "{{ foo | first | ['key'] }}" or "{{ foo | first | .key }}" or "{{ foo | first | get('key') }}".

One way to accomplish this is to "map" out the key across all the results and then limit to the first item. Here's an example that should be a drop in replacement for your code:

```
  - debug:
      msg: "{{ copy_jars.results | map(attribute='dest') | first }}"
```

To sum up, this is not a bug in ansible or in jinja, but is really a confusion about using the attr() filter.

**Comment by jhawkesworth** [comment]
Thanks for the explanation @jctanner - I had hit exactly the same thing in a template early today and came up with the same workaround.

**Comment by GoGoris** [comment]
You also have my thanks @jctanner! I am sure that it will save people with the same issue a lot of time.

**Comment by hlm** [comment]
I had the similar problem, but trying to get the `last` item. 

This worked for me:

```
  - debug:
      msg: "{{ (copy_jars.results | last)["dest"] }}"
```

Interestingly, the approach based on `map` did not work for me while using last. I did not dig too much into the details, but it complained about `reverse()` requiring a sequence (shown here just for reference):

```
{"failed": true, "msg": "Unexpected templating type error occurred on ({{adds.results|map(attribute='taken')|last|length}}): argument to reversed() must be a sequence"}```
